Resource publish tuning

remuslazar opened the pull request:
This PR fixes the issue with `flow resource:publish` freaking out by using all available memory and not doing anything.

This PR changes the underlying logic by using the Doctrine Iterator and Query directly by calling the \Iterable API instead of dealing with large internal arrays.

Additionally this introduces a new `Resource.lastPublishedDateTime` field which saves the timestamp when the resource was last published (using the command controller). The `resource:publish` CommandController accepts a timestamp as an option and, if set will limit the processing to the records which are "newer than the given timestamp OR do not have the timestamp field set at all (=null).

remuslazar committed:
Performance Tuning for the resource:publish call (WIP)

use the \Traversable object instead of dealing with arrays

baschny commented on TYPO3.Flow/Classes/TYPO3/Flow/Resource/Target/FileSystemTarget.php:
```diff
@@ -156,7 +157,15 @@ public function getName()
     public function publishCollection(Collection $collection)
     {
-        foreach ($collection->getObjects() as $object) {
+        foreach ($collection->findResources() as $resource) {
```
@remuslazar aha, I see - du bist schon am optimieren... Bin gespannt!

remuslazar committed:
Resource:publish progress indicator implemented

remuslazar committed:
Use the \Iterable interface methods for traversal instead of foreach()

remuslazar committed:
Use the doctrine ORM query object for hydration

the Flow query object does materialize the result as an array, so we
do use the underlying Doctrine Query directly to avoid that.

remuslazar committed:
use Doctrine Query ->iterate() instead of execute()

remuslazar committed:
Don't bail out if a single resource cannot be published

this should be refined..

remuslazar committed:
detach object from Doctrine ORM so it can be GC'd

remuslazar commented:
@baschny das funktioniert nun, Memory steigt jedoch geringfügig an (das Logging hat vermutlich auch irgendwelche Caches oder Objekte, die nicht GC'd werden..), mit 4GB limit geht das jedoch durch, bei ca. 70% hat der PHP Prozess knapp 1.5GB verbraucht.

Daher schlage ich vor, dass wir das mergen.

### Upstream PR

Upstream PR mache ich jetzt erst mal nicht, das können wir immer noch später machen, nach dem Upgrade. Dann kann ich das gerne hoch ziehen und es ggf. auch "in Schön" machen. Mache das jetzt aber nicht aus pragmatischen Gründen.

remuslazar committed:
Flow\Resource.lastPublishedDateTime field and migration added

remuslazar committed:
Collection.publish() signature changes (make the progressFn param optional)

remuslazar committed:
Resource.publish, set setLastPublishedDateTime to now()

remuslazar committed:
Resource.publish, setLastPublishedDateTime logic tweaks

Do set the timestamp only while running the batch resource:publish

remuslazar committed:
resource:publish newerThan handling implemented

remuslazar commented:
@baschny last published field und logic wäre nun implementiert. Ich habe das schon auf AWS deployed und das "nudelt" noch im screen mit time davor. Schauen wir mal, wie lange es braucht.

Dieses "last modified" ist auch ganz nützlich wenn es aus irgendeinem Grunde abbricht. So kann man damit auch ein "resume" machen.. muss man nicht die noch durchgehen, wie man schon "gepublished" hat.

Wenn das soweit tut, würde ich folgendes tun:

1. dieses PR mergen und auf LIVE (aktuell) deployen
1. Setting `subdivideHashPathSegment` vornehmen
1. _kein_ cache:flush machen für `Production\Live`
1. `export FLOW_CONTEXT=Production\Live\Temp`
1. `resource:publish` machen
1. `export FLOW_CONTEXT=Production\Live`
1. `flow:cache:flush` machen

Eventuell kann man mit einer nginx rule das Hashing berücksichtigen und dann alles mit 
`find -maxdepth 1 -type d  -name '????????????????????????????????????????' -exec rm -rf {} \;` löschen.

remuslazar commented:
## DAZ Live Datenbestand resource:publish

### t2.medium

```
[www_live@ip-10-140-1-62 www_live]$ date ; time ./flow resource:publish ; date
Fr 5. Okt 16:22:20 UTC 2018
Publishing resources of collection "static"
    0 [============================]
Publishing resources of collection "persistent"
 683935 [============================]

real    470m51.656s
user    14m54.966s
sys     4m26.994s
Sa 6. Okt 00:13:12 UTC 2018
```

Das Problem war wohl swap und IO Credits, da 4GB für diese Operation knapp bemessen waren (Redis und ES haben auch schon einiges am RAM verbraucht, daher blieb nicht viel übrig für PHP). Ich lasse das gerade auf t2.large (8GB anstatt 4GB RAM) laufen um die Werte hinterher vergleichen zu können.

### t2.large

```
Sa 6. Okt 06:31:30 UTC 2018
Publishing resources of collection "static"
    0 [============================]
Publishing resources of collection "persistent"
 683935 [============================]

real    75m33.696s
user    11m47.254s
sys     2m6.273s
Sa 6. Okt 07:47:04 UTC 2018
```

remuslazar committed:
Disable Doctrine ORM result cache

should be disabled by default but we want to be explicit, though

remuslazar committed:
Disable logging while running resource:publish

remuslazar commented:
@baschny also mit 8GB RAM läuft das in 75 Minuten durch. Ich habe jetzt noch 2 Optimierungen vorgenommen (kein Logging und disable result cache), mal sehen, wie sich diese auswirken. Lasse das zum testen nochmals laufen in der AWS Umgebung.

remuslazar committed:
Batch processing performance tuning (resource pool)

use a resource pool to speedup execution of SQL updates. The pool will
add up to 500 entities and update them as a single batch of changes.
Using that we can heavily reduce the load on the SQL server.

remuslazar commented:
@baschny so, habe es noch weiter optimiert, siehe 3e59fe0. Nun sieht das so aus:

```
[www_live@ip-10-140-1-62 www_live]$ date ; time ./flow resource:publish ; date
Sa 6. Okt 09:20:24 UTC 2018
Publishing resources of collection "static"
    0 [============================]
Publishing resources of collection "persistent"
 683935 [============================]

real    26m58.892s
user    10m55.222s
sys     1m29.635s
Sa 6. Okt 09:47:23 UTC 2018
```

remuslazar committed:
resource:clean: use the Doctrine iterate() API if available

remuslazar commented:
@baschny Ich habe nun auch `resource:clean` refactored, damit das durch geht. Sieht nun so aus:

```
[www_live@ip-10-140-1-62 www_live]$ ./flow resource:clean
Checking if resource data exists for all known resource objects ...

 683935/683935 [============================] 100%

Found 14 broken resource(s):

Bepanthen nasensalben dpa-580x307.jpg (06426a6c822319442168bd4e5ffe6898636fcc06) persistent
Bepanthen nasensalben dpa-1000x529.jpg (463f59aecd506af534a53933a2452809a4f9d8cd) persistent
Bepanthen nasensalben dpa.jpg (9693f938136941c8833ef1ecaf388fbf992aaba9) persistent
 -> TYPO3\Media\Domain\Model\Image (abbf95bd-5981-9379-6beb-e48dcea8e1b9)
Zytos benicoma AdobeStock_106335805-6016x3375.jpeg (1b3fad5cee3c90604aa6ff4d61974c7cf6a24566) persistent
 -> TYPO3\Media\Domain\Model\ImageVariant (6ea4f618-09b2-9135-748a-f404e2c59098)
IMG_1522 (1)-640x853.jpg (90644c1dc9fabc9107e0f900a33647c3ff4734d1) persistent
 -> TYPO3\Media\Domain\Model\ImageVariant (3076554e-781f-a46e-5231-eddd1b0e694e)
Kirsten Sucker-1043x696-112x75.jpg (bfd985bde5b955b7515651329a6ef9d7e7130bb4) persistent
Homöopathie Globuli Holztisch Apotheke___bina01__AdobeStock_96220071-3500x1995-637x363.jpg (9cbfc1b91553de0ee961d914d20b5b91877b5967) persistent
20160317_114357-1665x880-301x159.jpg (c5a61de2b010314f8fa5e9f0a286b3fbfb4924ec) persistent
Bepanthen nasensalben dpa-250x132.jpg (a713aa9c36b9a5f472ad3d9c0c4607d071ff02a0) persistent
Bepanthen nasensalben dpa-600x317.jpg (2d84c6d3749446929ec58cc014d0eca6339fa978) persistent
IMG_1522 (1)-640x853-225x300.jpg (6d17788d4f997ccce89037296263b1acbeb20d87) persistent
20160317_114357-1665x880.jpg (57c5d1c6aef06ca391a763ae80fedf239477e677) persistent
 -> TYPO3\Media\Domain\Model\ImageVariant (01d79345-89fd-1e3f-dc1f-7c9953588fc9)
Bepanthen nasensalben dpa-40x40.jpg (4ab55cc66ba28a2f0f9e641398d45f7c9f47f009) persistent
Kirsten Sucker-1043x696.jpg (b305393f8948e81e99f9dd6c8a0501a3195415ba) persistent
 -> TYPO3\Media\Domain\Model\ImageVariant (6787bd17-3545-abb4-38b4-9eb904b8a0dc)
Do you want to remove all broken resource objects and related assets from the database? (y/n/c) n
Did not delete any resource objects.
```

baschny commented:
Wir machen die Änderungen in der Konfiguration (Hash-Change) erst beim Live-Going der AWS Umgebung, d.h. wir Deployen oder Ändern erstmal nichts weiteres in der aktuellen Live-Umgebung.
